src/python/pants/backend/jvm/tasks/coverage/scoverage.py

```diff
@@ -0,0 +1,152 @@
+
+import functools
+import os
+import subprocess
+import re
+from pants.java.jar.jar_dependency import JarDependency
+from pants.backend.jvm.tasks.coverage.engine import CoverageEngine
+from pants.base.exceptions import TaskError
+from pants.subsystem.subsystem import Subsystem
+from pants.util.dirutil import relativize_paths, safe_mkdir, safe_mkdir_for, safe_walk, touch
+
+
+report_generator = 'src/scala/org/pantsbuild/scoverage/report:gen2'
+
+
+class Scoverage(CoverageEngine):
+  """Class to run coverage tests with scoverage"""
+
+
+  class Factory(Subsystem):
+    options_scope = 'scoverage-runtime'
+
+    @classmethod
+    def create(cls, settings, targets, execute_java_for_targets):
+      """
+      :param settings: Generic code coverage settings.
+      :type settings: :class:`CodeCoverageSettings`
+      :param list targets: A list of targets to instrument and record code coverage for.
+      :param execute_java_for_targets: A function that accepts a list of targets whose JVM platform
+                                       constraints are used to pick a JVM `Distribution`. The
+                                       function should also accept `*args` and `**kwargs` compatible
+                                       with the remaining parameters accepted by
+                                       `pants.java.util.execute_java`.
+      """
+
+      return Scoverage(settings, targets, execute_java_for_targets)
+
+
+  @staticmethod
+  def register_junit_options(register, register_jvm_tool):
+
+    def scoverage_runtime_jar(**kwargs):
+      return JarDependency(org='com.twitter.scoverage', name='scalac-scoverage-runtime_2.12',
+        rev='1.0.1-twitter', **kwargs)
+
+    register_jvm_tool(register,
+      'scalac-scoverage-runtime',
+      classpath=[
+        scoverage_runtime_jar()
+      ])
+
+  def __init__(self, settings, targets, execute_java_for_targets):
+    """
+    :param settings: Generic code coverage settings.
+    :type settings: :class:`CodeCoverageSettings`
+    :param list targets: A list of targets to instrument and record code coverage for.
+    :param execute_java_for_targets: A function that accepts a list of targets whose JVM platform
+                                     constraints are used to pick a JVM `Distribution`. The function
+                                     should also accept `*args` and `**kwargs` compatible with the
+                                     remaining parameters accepted by
+                                     `pants.java.util.execute_java`.
+    """
+    self._settings = settings
+    self._context = settings.context
+    self._targets = targets
+    self._target_filters = []    # TODO(sameera): add target filtering for scoverage
+    self._execute_java = functools.partial(execute_java_for_targets, targets)
+
+  def _iter_datafiles(self, output_dir):
+    for root, _, files in safe_walk(output_dir):
+      for f in files:
+        if f.startswith("scoverage"):
```

Comment:
Where does the hard-coded scoverage path come from?

Reply:
So all the scoverage instrument files have the name "scoverage.coverage". Similarly, measurements start with the name "scoverage.measurements.<TID>". Here I clean up the measurements dir by removing all the files beginning with the name "scoverage". Thanks for pointing it out (mentioned it as a comment on the code)

Reply:
Should move these comments into python method docs:
```
def _iter_datafiles(self, output_dir):
   """This is a docstring for a method.

   With some other stuff in it.
   """
   ..
```